SafeHaskell not working

Given:

`ghcide-safe.cabal`:
```
cabal-version:       >=1.10
name:                ghcide-safe
version:             0.1.0.0
build-type:          Simple

library
  exposed-module:      Safe
                       TrustWorthy
  build-depends:       base >=4.12 && <4.13
  default-language:    Haskell2010
```

`TrustWorthy.hs`
```haskell
{-# LANGUAGE Trustworthy #-}
module TrustWorthy where

import System.IO.Unsafe
import System.IO

trustWorthyId :: a -> a
trustWorthyId i = unsafePerformIO $ do
  putStrLn "I'm safe"
  return i
```

`Safe.hs`
```haskell
{-# LANGUAGE Safe #-}
module Safe where

import TrustWorthy

safeId :: a -> a
safeId = trustWorthyId
```

Running `ghcide` gives me:

```
$ ghcide
ghcide version: 0.1.0 (GHC: 8.6.5) (PATH: /home/christiaan/.cabal/store/ghc-8.6.5/ghcide-0.1.0-9dff19bbe1a6479cf0636798849c8610d18a87b60e9dd627dabbfb830d06787a/bin/ghcide)
Ghcide setup tester in /home/christiaan/devel/ghcide-safe.
Report bugs at https://github.com/digital-asset/ghcide/issues

Step 1/6: Finding files to test in /home/christiaan/devel/ghcide-safe
Found 3 files

Step 2/6: Looking for hie.yaml files that control setup
Found 1 cradle

Step 3/6, Cradle 1/1: Implicit cradle for /home/christiaan/devel/ghcide-safe
Cradle {cradleRootDir = "/home/christiaan/devel/ghcide-safe", cradleOptsProg = CradleAction: Default}

Step 4/6, Cradle 1/1: Loading GHC Session
Loaded package environment from /home/christiaan/devel/ghcide-safe/.ghc.environment.x86_64-linux-8.6.5

Step 5/6: Initializing the IDE

Step 6/6: Type checking the files
File:     /home/christiaan/devel/ghcide-safe/Safe.hs
Hidden:   no
Range:    4:0-4:19
Source:   typecheck
Severity: DsError
Message: 
  /home/christiaan/devel/ghcide-safe/Safe.hs:4:1: error:
  TrustWorthy: Can't be safely imported!
  The module itself isn't safe.
Files that failed:
 * /home/christiaan/devel/ghcide-safe/Safe.hs

Completed (2 files worked, 1 file failed)
```

--- a/src/Development/IDE/Core/Compile.hs
+++ b/src/Development/IDE/Core/Compile.hs
@@ -234,10 +234,11 @@ mkTcModuleResult
     -> m TcModuleResult
 mkTcModuleResult tcm = do
     session <- getSession
+    let sf = modInfoSafe (tm_checked_module_info tcm)
 #if MIN_GHC_API_VERSION(8,10,0)
-    iface <- liftIO $ mkIfaceTc session Sf_None details tcGblEnv
+    iface <- liftIO $ mkIfaceTc session sf details tcGblEnv
 #else
-    (iface, _) <- liftIO $ mkIfaceTc session Nothing Sf_None details tcGblEnv
+    (iface, _) <- liftIO $ mkIfaceTc session Nothing sf details tcGblEnv
 #endif
     let mod_info = HomeModInfo iface details Nothing
     return $ TcModuleResult tcm mod_info

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -63,6 +63,7 @@ main = defaultMainWithRerun $ testGroup "HIE"
   , pluginTests
   , preprocessorTests
   , thTests
+  , safeTests
   , unitTests
   , haddockTests
   , positionMappingTests
@@ -1484,6 +1485,36 @@ preprocessorTests = testSessionWait "preprocessor" $ do
         [(DsError, (2, 8), "Variable not in scope: z")]
       )
     ]
+
+
+safeTests :: TestTree
+safeTests =
+  testGroup
+    "SafeHaskell"
+    [ -- Test for https://github.com/digital-asset/ghcide/issues/424
+      testSessionWait "load" $ do
+        let sourceA =
+              T.unlines
+                ["{-# LANGUAGE Trustworthy #-}"
+                ,"module A where"
+                ,"import System.IO.Unsafe"
+                ,"import System.IO"
+                ,"trustWorthyId :: a -> a"
+                ,"trustWorthyId i = unsafePerformIO $ do"
+                ,"  putStrLn \"I'm safe\""
+                ,"  return i"]
+            sourceB =
+              T.unlines
+                ["{-# LANGUAGE Safe #-}"
+                ,"module B where"
+                ,"import A"
+                ,"safeId :: a -> a"
+                ,"safeId = trustWorthyId"
+                ]
+
+        _ <- openDoc' "A.hs" "haskell" sourceA
+        _ <- openDoc' "B.hs" "haskell" sourceB
+        expectNoMoreDiagnostics 1 ]
 
 thTests :: TestTree
 thTests =

